Adding troubleshooting section (#1423)
pyansys/pymapdl: doc/source/troubleshoot/troubleshoot.rst

    *** ERROR ***
    Another Ansys job with the same job name (file) is already running in this
    directory or the file.lock file has not been deleted from an abnormally
    terminated Ansys run. To disable this check, set the ANSYS_LOCK environment
    variable to OFF.

There are many issues that can cause Ansys not to launch, including:

- License server setup
- Running behind a VPN
- Missing dependencies
- Conflicts with a student version

Licensing Issues
================

Incorrect license server configuration can prevent Ansys from being able to get a valid license.
In those cases, you might see output **similar** to:

.. code:: pwsh

   (base) PS C:\Users\user\temp> & 'C:\Program Files\ANSYS Inc\v222\ansys\bin\winx64\ANSYS222.exe'

   ANSYS LICENSE MANAGER ERROR:

   Maximum licensed number of demo users already reached.

   ANSYS LICENSE MANAGER ERROR:

   Request name mech_2 does not exist in the licensing pool.
   No such feature exists.
   Feature:          mech_2
   License path:  C:\Users\user\AppData\Local\Temp\\cb0400ba-6edb-4bb9-a333-41e7318c007d;
   FlexNet Licensing error:-5,357

> PADT generally has a great blog regarding ANSYS issues, and licensing is always a common issue 
> (for example `Changes to Licensing at ANSYS 2020R1 <https://www.padtinc.com/blog/15271-2/>`_).  
> Should you be responsible for maintaining Ansys licensing or have a personal install of Ansys,
> please check the online Ansys licensing documentation at 
> `Installation and Licensing <https://ansyshelp.ansys.com/account/secured?returnurl=/Views/Secured/prod_page.html?pn=Installation%20and%20Licensing&pid=InstallationAndLicensing&lang=en>`_.

Reviewer: ```suggestion
PADT has a great blog regarding ANSYS issues, and licensing is always a common issue 
(for example `Changes to Licensing at ANSYS 2020R1 <https://www.padtinc.com/blog/15271-2/>`_).  
Should you be responsible for maintaining Ansys licensing or have a personal install of Ansys,
please check the online Ansys licensing documentation at 
`Installation and Licensing <https://ansyshelp.ansys.com/account/secured?returnurl=/Views/Secured/prod_page.html?pn=Installation%20and%20Licensing&pid=InstallationAndLicensing&lang=en>`_.
```

using Ubuntu 16.04, you can install that simply with ``sudo apt
install libxp6``.  However, on Ubuntu 18.04+, you must manually
download and install the package.

Since ``libxpl6`` also pre-depends on ``multiarch-support``, which is
also outdated, it must be removed, otherwise you'll have a broken
package configuration.  The following step downloads and modifies the
``libxp6`` package to remove the ``multiarch-support`` dependency, and
then installs it via ``dpkg``.

.. code::

    cd /tmp
    wget http://ftp.br.debian.org/debian/pool/main/libx/libxp/libxp6_1.0.2-2_amd64.deb
    ar x libxp6_1.0.2-2_amd64.deb
    sudo tar xzf control.tar.gz
    sudo sed '/Pre-Depends/d' control -i
    sudo bash -c "tar c postinst postrm md5sums control | gzip -c > control.tar.gz"
    sudo ar rcs libxp6_1.0.2-2_amd64_mod.deb debian-binary control.tar.gz data.tar.xz
    sudo dpkg -i ./libxp6_1.0.2-2_amd64_mod.deb

.. _conflicts_student_version:

Conflicts with Student Version
==============================

Although you can install Ansys together with other Ansys products or versions, on Windows, you
should not install a student version of an Ansys product together with its non-student version.

> For example, installing both the Ansys MAPDL 202 2R2 Student Version and Ansys MAPDL 2022

Reviewer: ```suggestion
For example, installing both the Ansys MAPDL202 2R2 Student Version and Ansys MAPDL 2022
```

    PS echo $env:AWP_ROOT222
    C:\Program Files\ANSYS Inc\v222

    PS echo $env:ANSYS222_DIR
    C:\Program Files\ANSYS Inc\ANSYS Student\v222\ANSYS
    PS [System.Environment]::SetEnvironmentVariable('ANSYS222_DIR','C:\Program Files\ANSYS Inc\v222\ANSYS',[System.EnvironmentVariableTarget]::User)
    PS echo $env:ANSYS222_DIR
    C:\Program Files\ANSYS Inc\v222\ANSYS

    PS echo $env:CADOE_LIBDIR222
    C:\Program Files\ANSYS Inc\ANSYS Student\v222\CommonFiles\Language\en-us
    PS [System.Environment]::SetEnvironmentVariable('CADOE_LIBDIR222','C:\Program Files\ANSYS Inc\v222\CommonFiles\Language\en-us',[System.EnvironmentVariableTarget]::User)
    PS echo $env:CADOE_LIBDIR222
    C:\Program Files\ANSYS Inc\v222\CommonFiles\Language\en-us

*****************
Launching PyMAPDL
*****************

Even if you are able to correctly launch MAPDL, PyMAPDL might have some problems to launch
MAPDL.

Manually Set the Executable Location
====================================
If you have a non-standard install, ``pymapdl`` may be unable find
your installation.  If that's the case, provide the location of MAPDL

> as the first parameter to ``launch_mapdl``.  For example, on Windows,

Reviewer: ```suggestion
as the first parameter to :func:`launch_mapdl <ansys.mapdl.core.launch_mapdl>`.  For example, on Windows,
```

this will be:

.. code:: python

    >>> from ansys.mapdl.core import launch_mapdl
    >>> exec_loc = 'C:/Program Files/ANSYS Inc/v211/ansys/bin/winx64/ANSYS211.exe'
    >>> mapdl = launch_mapdl(exec_loc)

For Linux:

.. code:: python

    >>> from ansys.mapdl.core import launch_mapdl
    >>> exec_loc = '/usr/ansys_inc/v211/ansys/bin/ansys211'
    >>> mapdl = launch_mapdl(exec_loc)

Should this fail to launch or hang while launching, pass

> ``verbose_mapdl=True`` when using ``launch_mapdl``.  This will print

Reviewer: ```suggestion
``verbose_mapdl=True`` when using :func:`launch_mapdl() <ansys.mapdl.core.launch_mapdl>`.  This will print
```

the output of MAPDL within Python and can be used to debug why MAPDL
isn't launching. On Windows, output is limited due to the way
MAPDL launches.

Default Executable Location
===========================

The first time that you run PyMAPDL, it detects the

> available ANSYS installations.

Reviewer: ```suggestion
available Ansys installations.
```

On Windows, Ansys installations are normally under:

.. code:: text

    C:/Program Files/ANSYS Inc/vXXX

On Linux, Ansys installations are normally under:

.. code:: text

    /usr/ansys_inc/vXXX

> If PyMAPDL finds a valid ANSYS installation, it caches its

Reviewer: ```suggestion
If PyMAPDL finds a valid Ansys installation, it caches its
```

path in the configuration file, ``config.txt``, whose path is shown in the
following code:

.. code:: python

    >>> from ansys.mapdl.core.launcher import CONFIG_FILE
    >>> print(CONFIG_FILE)
    'C:\\Users\\user\\AppData\\Local\\ansys_mapdl_core\\ansys_mapdl_core\\config.txt'

In certain cases, this configuration might become obsolete. For example, when a new
Ansys version is installed and an earlier installation is removed.
To update this configuration file with the latest path, use:

.. code:: python

    >>> from ansys.mapdl.core import save_ansys_path
    >>> save_ansys_path(r"C:\Program Files\ANSYS Inc\v222\ansys\bin\winx64\ansys222.exe")
    'C:\\Program Files\\ANSYS Inc\\v222\\ansys\\bin\\winx64\\ansys222.exe'

If you want to check which Ansys installations PyMAPDL has detected, use:

.. code:: python

>     >>> from ansys.mapdl.core.launcher import _get_available_base_ansys

Reviewer: Consider making this a public function if you include this in the docs.

Author: Fair point.

    >>> _get_available_base_ansys()
    {222: 'C:\\Program Files\\ANSYS Inc\\v222',
    212: 'C:\\Program Files\\ANSYS Inc\\v212',

>     -222: 'C:\\Program Files\\ANSYS Inc\\ANSYS Student\\v222'}

Reviewer: ```suggestion
    222: 'C:\\Program Files\\ANSYS Inc\\ANSYS Student\\v222'}
```

>     {222: 'C:\\Program Files\\ANSYS Inc\\v222',
>     212: 'C:\\Program Files\\ANSYS Inc\\v212',
>     -222: 'C:\\Program Files\\ANSYS Inc\\ANSYS Student\\v222'}

Reviewer: ```suggestion
    {222: 'C:\\Program Files\\ANSYS Inc\\v222',
     212: 'C:\\Program Files\\ANSYS Inc\\v212',
     222: 'C:\\Program Files\\ANSYS Inc\\ANSYS Student\\v222'}
```

Author: This regarding this.

Reviewer: Ah, cool. Could you make a note of that?

Author: Yes, at the function and at the docs.

Student versions are provided as **negative** versions because the Python dictionary
does not accept two equal keys. The result of the
:func:`get_available_ansys_installations() <ansys.mapdl.core.get_available_ansys_installations>`

> method lists higher versions first and student versions last.

Author: Here @akaszynski

are several ways to improve performance and stability of MADPL:

- When possible, pass ``mute=True`` to individual MAPDL commands or
  set it globally with :func:`Mapdl.mute
  <ansys.mapdl.core.mapdl_grpc.MapdlGrpc>`.  This disables streaming
  back the response from MAPDL for each command and will marginally
  improve performance and stability.  Consider having a debug flag in
  your program or script so you can enable or disable logging and
  verbosity when needed.

Issues
======

.. note::
   MAPDL 2021R1 has a stability issue with :func:`Mapdl.input()
   <ansys.mapdl.core.Mapdl.input>`.  Avoid using input files if
   possible.  Attempt to :func:`Mapdl.upload()
   <ansys.mapdl.core.Mapdl.upload>` nodes and elements and read them
   in via :func:`Mapdl.nread() <ansys.mapdl.core.Mapdl.nread>` and
   :func:`Mapdl.eread() <ansys.mapdl.core.Mapdl.eread>`.

.. _ref_pymapdl_limitations:

*******************
PyMAPDL Limitations
*******************

.. _ref_numpy_arrays_in_mapdl:

Issues when Importing and Exporting Numpy Arrays in MAPDL
=========================================================

Because of the way MAPDL is designed, there is no way to store an

> array where one or more dimension is zero.

Reviewer: ```suggestion
array where one or more dimensions are zero.
```

This can happens in Numpy arrays, where its first dimension can be

> set to zero.

Reviewer: ```suggestion
set to zero. For example:
```

.. code:: python

   >>> import numpy
   >>> from ansys.mapdl.core import launch_mapdl
   >>> mapdl = launch_mapdl()

>    >>> array40 = np.reshape([1, 2, 3, 4], (4,))
>    >>> array40

Reviewer: ```suggestion
   >>> my_array = np.reshape([1, 2, 3, 4], (4,))
   >>> my_array
```

   array([1, 2, 3, 4])

These types of array dimensions will be always converted to ``1``.
For example:

.. code:: python

   >>> mapdl.parameters['mapdlarray40'] = array40
   >>> mapdl.parameters['mapdlarray40']
   array([[1.],
      [2.],
      [3.],
      [4.]])
   >>> mapdl.parameters['mapdlarray40'].shape
   (4, 1)

> This means that when you pass two arrays, one with the second axis equal
> to zero (e.g. ``array40``) and another one with the second axis equal
> to one, if later retrieved, they will have the same
> shape.
> 
> .. code:: python
> 
>    >>> array41 = np.reshape([1, 2, 3, 4], (4,1))
>    >>> array41
>    array([[1],
>       [2],
>       [3],
>       [4]])
>    >>> mapdl.parameters['mapdlarray41'] = array41
>    >>> mapdl.parameters['mapdlarray41']
>    array([[1.],
>       [2.],
>       [3.],
>       [4.]])
>    >>> np.allclose(mapdl.parameters['mapdlarray40'], mapdl.parameters['mapdlarray41'])
>    True

Reviewer: ```suggestion
This means that when you pass two arrays, one with the second axis equal
to zero (e.g. ``my_array``) and another one with the second axis equal
to one, if later retrieved, they will have the same shape.

.. code:: python

   >>> my_other_array = np.reshape([1, 2, 3, 4], (4,1))
   >>> my_other_array
   array([[1],
      [2],
      [3],
      [4]])
   >>> mapdl.parameters['mapdlarray_b'] = my_other_array
   >>> mapdl.parameters['mapdlarray_b']
   array([[1.],
      [2.],
      [3.],
      [4.]])
   >>> np.allclose(mapdl.parameters['mapdlarray'], mapdl.parameters['mapdlarray_b'])
   True
```
